pkg/controllers/staticroute/staticroute_controller.go

```diff
@@ -72,60 +71,86 @@ func deleteSuccess(r *StaticRouteReconciler, _ context.Context, o *v1alpha1.Stat
+	return ResultNormal, nil
+
+}
 func (r *StaticRouteReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	obj := &v1alpha1.StaticRoute{}
 	log.Info("reconciling staticroute CR", "staticroute", req.NamespacedName)
 	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerSyncTotal, common.MetricResTypeStaticRoute)
 
 	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
-		log.Error(err, "unable to fetch static route CR", "req", req.NamespacedName)
-		return ResultNormal, client.IgnoreNotFound(err)
-	}
-
-	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
-		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateTotal, common.MetricResTypeStaticRoute)
-		if !controllerutil.ContainsFinalizer(obj, commonservice.StaticRouteFinalizerName) {
-			controllerutil.AddFinalizer(obj, commonservice.StaticRouteFinalizerName)
-			if err := r.Client.Update(ctx, obj); err != nil {
-				log.Error(err, "add finalizer", "staticroute", req.NamespacedName)
-				updateFail(r, ctx, obj, &err)
+		if client.IgnoreNotFound(err) == nil {
+			if err := r.deleteStaticRouteByName(req.Namespace, req.Name); err != nil {
 				return ResultRequeue, err
+			} else {
+				return ResultNormal, nil
 			}
-			log.V(1).Info("added finalizer on staticroute CR", "staticroute", req.NamespacedName)
 		}
+		log.Error(err, "unable to fetch static route CR", "req", req.NamespacedName)
+		return ResultRequeue, err
+	}
 
-		if err := r.Service.CreateOrUpdateStaticRoute(req.Namespace, obj); err != nil {
-			updateFail(r, ctx, obj, &err)
-			// TODO: if error is not retriable, not requeue
-			apierror, errortype := util.DumpAPIError(err)
-			if apierror != nil {
-				log.Info("create or update static route failed", "error", apierror, "error type", errortype)
-			}
-			return ResultRequeue, err
-		}
-		updateSuccess(r, ctx, obj)
-	} else {
-		if controllerutil.ContainsFinalizer(obj, commonservice.StaticRouteFinalizerName) {
-			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteTotal, common.MetricResTypeStaticRoute)
-			// TODO, update the value from 'default' to actual value， get OrgID, ProjectID, VPCID depending on obj.Namespace from vpc store
-			if err := r.Service.DeleteStaticRoute(obj); err != nil {
-				log.Error(err, "delete failed, would retry exponentially", "staticroute", req.NamespacedName)
-				deleteFail(r, ctx, obj, &err)
-				return ResultRequeue, err
-			}
-			controllerutil.RemoveFinalizer(obj, commonservice.StaticRouteFinalizerName)
-			if err := r.Client.Update(ctx, obj); err != nil {
-				deleteFail(r, ctx, obj, &err)
-				return ResultRequeue, err
-			}
-			log.V(1).Info("removed finalizer", "staticroute", req.NamespacedName)
-			deleteSuccess(r, ctx, obj)
-		} else {
-			// only print a message because it's not a normal case
-			log.Info("finalizers cannot be recognized", "staticroute", req.NamespacedName)
-		}
+	if !obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.deleteStaticRoute(ctx, req, obj)
```

Comment:
For this case, you can get CR when deleting, so you can keep logic
if err := r.Service.DeleteStaticRoute(obj); err != nil {
				log.Error(err, "delete failed, would retry exponentially", "staticroute", req.NamespacedName)
				deleteFail(r, ctx, obj, &err)
				return ResultRequeue, err
			}
			deleteSuccess(r, ctx, obj)